slo-monitor/src/monitors/store.go

```diff
@@ -0,0 +1,208 @@
+	// we also store the offset (explained below).
+	indices map[string]int64
+	// index that will be assigned to next item that'll be added to the queue
+	nextIndex int64
+	// index of the (last process item + 1) which is equivalent to the index of the first item currently in the queue.
+	// This means that indices in the queue are [offset, offset+1, offset+2, ...].
+	offset int64
+}
+
+var keyFunc = cache.DeletionHandlingMetaNamespaceKeyFunc
+
+// NewNoStoreQueue returns newly created NoStoreQueue
+func NewNoStoreQueue() *NoStoreQueue {
+	q := &NoStoreQueue{
+		queue:   []workItem{},
+		indices: map[string]int64{},
+	}
+
+	q.cond.L = q
+	return q
+}
+
+func (q *NoStoreQueue) enqueueOrModifyItemLocked(key string, obj interface{}, operationType string) {
+	if index, ok := q.indices[key]; ok {
+		q.queue[int(int64(index)-q.offset)].operationType = operationType
+		q.queue[int(int64(index)-q.offset)].obj = obj
+		return
+	}
+	q.indices[key] = q.nextIndex
+	q.nextIndex++
+	q.queue = append(q.queue, workItem{
+		operationType: operationType,
+		key:           key,
+		obj:           obj,
+	})
+}
+
+// Add implements Add function from Store interface
+func (q *NoStoreQueue) Add(obj interface{}) error {
+	key, err := keyFunc(obj)
+	if err != nil {
+		return cache.KeyError{obj, err}
+	}
+
+	q.Lock()
+	defer q.cond.Broadcast()
+	defer q.Unlock()
+	q.enqueueOrModifyItemLocked(key, obj, setOperationType)
+	return nil
+}
+
+// Update implements Update function from Store interface
+func (q *NoStoreQueue) Update(obj interface{}) error {
+	return q.Add(obj)
+}
+
+// Delete implements Delete function from Store interface
+func (q *NoStoreQueue) Delete(obj interface{}) error {
+	key, err := keyFunc(obj)
+	if err != nil {
+		return cache.KeyError{obj, err}
+	}
+
+	q.Lock()
+	defer q.cond.Broadcast()
+	defer q.Unlock()
+	q.enqueueOrModifyItemLocked(key, obj, deleteOperationType)
+	return nil
+}
+
+// List implements List function from Store interface
+func (q *NoStoreQueue) List() []interface{} {
+	return []interface{}{}
+}
+
+// ListKeys implements ListKeys function from Store interface
+func (q *NoStoreQueue) ListKeys() []string {
+	return []string{}
+}
+
+// Get implements Get function from Store interface
+func (q *NoStoreQueue) Get(obj interface{}) (item interface{}, exists bool, err error) {
+	return "", false, fmt.Errorf("Unimplemented Get")
+}
+
+// GetByKey implements GetByKey function from Store interface
+func (q *NoStoreQueue) GetByKey(key string) (item interface{}, exists bool, err error) {
+	return "", false, fmt.Errorf("Unimplemented GetByKey")
+}
+
+// Replace implements Replace function from Store interface
+func (q *NoStoreQueue) Replace(list []interface{}, rv string) error {
+	// Replace is called during re-lists, so we can't return "unimplemented" here.
+	// It's safe to ignore this call, as worst thing that can happen is loosing
+	// few calls. As this is best effort component, we don't care too much about that.
+	return nil
```

Comment:
What do you mean by "return" - actually keep the store with all keys? This would be a pretty big change, and the possible memory leak most likely wouldn't be too big.